portal/src/main/webapp/js/src/patient-view/util/ClinicalAttributesUtil.js

```diff
@@ -0,0 +1,203 @@
+            if ($.isNumeric(value)) {
+              value = Math.round(value);
+            }
+            cleanClinicalData[key] = value;
+            break;
+        }
+      }
+    }
+    return cleanClinicalData;
+  }
+
+  /**
+   * Get first key found in object. Otherwise return null.
+   * @param {object} object - object with key/value pairs
+   * @param {array} keys - array of keys
+   */
+  function getFirstKeyFound(object, keys) {
+      if (!object) {
+        return null;
+      }
+
+      for (var i=0; i< keys.length; i++) {
+          var value = object[keys[i]];
+          if (typeof value !== 'undefined' && value !== null) {
+            return value;
+          }
+      }
+      return null;
+  }
+
+
+  /**
+   * Derive clinical attributes from existing clinical attributes .e.g. age based
+   * on a date of birth. TODO: Now only includes a funky hack to keep current
+   * derived clinical attributes working.
+   * @param {object} clinicalData - key/value pairs of clinical data
+   */
+  function derive(clinicalData) {
+    var derivedClinicalAttributes = $.extend({}, clinicalData);
+    
+    /**
+     * TODO: Pretty funky function to get a normalized case type. This should
+     * probably also be a clinical attribute with a restricted vocabulary. Once
+     * the database has been changed to include normalized case types, this
+     * function should be removed.
+     * @param {object} clinicalData - key/value pairs of clinical data
+     * @param {string} caseTypeAttrs - TUMOR_TYPE or SAMPLE_TYPE value to normalize
+     */
+    function normalizedCaseType(clinicalData, caseTypeAttrs) {
+      var caseTypeNormalized = null;
+      var caseType;
+      var caseTypeLower;
+      var attr;
+      var i;
+
+      for (i = 0; i < caseTypeAttrs.length; i++) {
+        caseType = clinicalData[caseTypeAttrs[i]];
+
+        if (!cbio.util.checkNullOrUndefined(caseType)) {
+          caseTypeLower = caseType.toLowerCase();
+
+          if (caseTypeLower.indexOf('metasta') >= 0) {
+            caseTypeNormalized = 'metastasis';
+          } else if (caseTypeLower.indexOf('recurr') >= 0) {
+            caseTypeNormalized = 'recurrence';
+          } else if (caseTypeLower.indexOf('progr') >= 0) {
+            caseTypeNormalized = 'progressed';
+          } else if (caseTypeLower.indexOf('prim') >= 0 || 
+                    caseTypeLower.indexOf('prim') >= 0) {
+            caseTypeNormalized = 'primary';
+          }
+          if (!cbio.util.checkNullOrUndefined(caseTypeNormalized)) {
+            break;
+          }
+        }
+      }
+
+      return caseTypeNormalized;
+    }
+
+    var caseTypeNormalized = normalizedCaseType(clinicalData, ['SAMPLE_TYPE', 'TUMOR_TISSUE_SITE', 'TUMOR_TYPE']);
+    if (caseTypeNormalized !== null) {
+      var loc;
+
+      derivedClinicalAttributes.DERIVED_NORMALIZED_CASE_TYPE = caseTypeNormalized;
+
+      // TODO: DERIVED_SAMPLE_LOCATION should probably be a clinical attribute.
```

Comment:
What do you mean by this comment? Is `TUMOR_SITE` a bad name for this attribute?